src/ParallelTempering.jl

```diff
@@ -144,55 +306,189 @@ function run!(replica::Replica, T_sched::Function;
+        end
+
+        if mcs % update_interval == 0
+            # replica flow 'f' and smoothed replica flow 'fs'
+            N_labeled = (replica.N_up + replica.N_down)    
+            f = ((N_labeled > 0) ? replica.N_down/N_labeled : 0.0)
+            fs = (1-w)*f + w*replica.rank/(replica.N_ranks-1)
+
+            # acceptance rate between replicas i and i+1
+            A = ((replica.rank == replica.N_ranks-1) ? 0.0 : rex_accepts[2]/replica.N_rex)
+
+            # print out temperatures, replica flow, and acceptance rates for each update
+            fname = @sprintf("FBOPT_%03d.dat", N_updates)
+            gather_print(
+                replica,
+                [1.0/replica.sampler.β, fs, A],
+                fname
+            )
+
+            # perform feedback optimization for temperatures
+            feedback_update!(replica, N_updates; w=w, dT′=dT′)
+
+            N_updates += 1
+            rex_accepts[1] = rex_accepts[2] = 0
+            replica.N_up = replica.N_down = 0
+            replica.N_rex  = 0
+
+            # reset label
+            if replica.rank == 0
+                replica.label = 1
+            elseif replica.rank == replica.N_ranks-1
+                replica.label = -replica.N_ranks
+            else
+                replica.label = replica.N_ranks + replica.rank+1
+            end
+        end
+    end
+
+    return MPI.Allgather([1.0/replica.sampler.β], MPI.COMM_WORLD)
+end
+
+"""
+Start a parallel tempering (PT) simulation that saves measured averages,
+histograms, minimal energies, and configurations.
+
+Run w/ the command "mpiexec -n [n_procs] julia --project [julia_script.jl]".
+"""
+function run_PT!(
+    replica::Replica, 
+    T_sched::Function; 
+    therm_mcs::Int64=1000, 
+    measure_interval::Int64=10, 
+    rex_interval::Int64=1, 
+    max_mcs::Int64=1_000_000, 
+    bin_size::Float64=1.0, 
+    print_hist::Bool=false, 
+    print_xyz_ranks::Vector{Int64}=Int64[]
+)
+    # set replica sampling β using temperature (β = 1.0/(kT)) 
+    T = T_sched(replica.rank+1, replica.N_ranks)
+    set_temp!(replica.sampler, T)
 
+    # set replica neighbor β's  
+    for i in 1:2
+        if replica.nn_ranks[i] != -1
+            replica.nn_βs[i] = 1.0 / T_sched(replica.nn_ranks[i]+1, replica.N_ranks)
+        end
+    end
+
+    # initialize energy and equilibrate replica to it's distribution
+    replica.sampler.E = energy(replica.sampler.system)
+    replica.sampler.M = sum(replica.sampler.system)
+
+    thermalize!(replica.sampler, therm_mcs)
+    
     replica.sampler.nsweeps = 1
 
     # manually include these basic measurements for now
-    M = zeros(Float64, 3)
-    U = 0.0
+    M  = 0.0
+    M2 = 0.0
+    X  = 0.0
+    U  = 0.0
     U2 = 0.0
-    C = 0.0
+    C  = 0.0
 
-    N_rex = 0
-    N_measure = 0
+    system_size = length(replica.sampler.system)
+
+    # record timeseries for minimum energies 
+    Emin = typemax(Float64)
+    mcs_Emin = Vector{Vector{Float64}}()
```

Comment:
This might be a micro-optimization that doesn't really matter, but maybe consider making this a `Vector{Tuple{Float64, Float64}}` (or `Vector{Tuple{Int, Float64}}`) and changing the line that pushes to it. Since the Tuple is constant size, I think this will avoid a ton of allocations.